### PR TITLE
chore: type hints for the connnection class

### PR DESCRIPTION
**Pull request by dimaqq**
Pull the Connection type hints from #1097

Only minor cleanup beyond that, no logical changes.

### CODE REVIEW

File: juju/client/connection.py [hunk 1]
```diff
@@ -117,28 +106,41 @@ class Connection:
...
         bakery_client=None,
-        max_frame_size=None,
+        max_frame_size: int | None = None,
         retries=3,
         retry_backoff=10,
-        specified_facades: Optional[
-            Dict[str, Dict[Literal["versions"], Sequence[int]]]
-        ] = None,
+        specified_facades: dict[str, dict[Literal["versions"], Sequence[int]]]
+        | None = None,
```
Comment (james-garner-canonical) [inline juju/client/connection.py]: Is this ruff's autoformatting? There must be a nicer way to format this haha

File: juju/client/connection.py [hunk 2]
```diff
@@ -534,7 +536,7 @@ async def _do_ping():
             log.debug("ping failed because of closed connection")
             pass
 
-    async def rpc(self, msg, encoder=None):
+    async def rpc(self, msg: dict, encoder=None) -> dict:
```
Comment (james-garner-canonical) [inline juju/client/connection.py]: Do you have enough info to parametrize `dict` here? Even as `str, Any` or `str, JSONData`.

(Does pyright complain about the unparametrized generic?)

File: juju/client/connection.py [hunk 3]
```diff
@@ -744,8 +746,13 @@ async def _try_endpoint(endpoint, cacert, delay):
             # only executed if inner loop's else did not continue
             # (i.e., inner loop did break due to successful connection)
             break
+        else:
+            # impossible, work around https://github.com/microsoft/pyright/issues/8791
+            assert False  # noqa: B011
+
```
Comment (james-garner-canonical) [inline juju/client/connection.py:769]: Ok so this is a `for ... else`, which I mentally substitute as `for ... nobreak`. Reading through the nested loops above, I see that we will either break or raise. So now I understand that this is basically to tell pyright that `result = await task` will always have been evaluated after we exit the loop

I think this is a bit hard to follow (more the fault of the original code than anything), maybe we could do something like this instead
```py
result: TaskResult | None = None
# nasty loops here
assert result is not None  # loop will raise if it fails to set result
```

or instead add `# pyright: ignore [reportPossiblyUnboundVariable]` where we refer to result, e.g.
```python
        self._ws = result[0]  # pyright: ignore [reportPossiblyUnboundVariable]
        self.addr = result[1]  # pyright: ignore [reportPossiblyUnboundVariable]
        self.endpoint = result[2]  # pyright: ignore [reportPossiblyUnboundVariable]
        self.cacert = result[3]  # pyright: ignore [reportPossiblyUnboundVariable]
```

File: juju/model.py [hunk 4]
```diff
@@ -257,7 +259,9 @@ def get_entity(self, entity_type, entity_id, history_index=-1, connected=True):
 class ModelEntity:
     """An object in the Model tree"""
 
-    def __init__(self, entity_id, model, history_index=-1, connected=True):
+    entity_id: str
+
+    def __init__(self, entity_id: str, model: Model, history_index=-1, connected=True):
```
Comment (james-garner-canonical) [inline juju/model.py]: Better to explicitly annotate `history_index` and `connected` as well if they're actually `int` and `bool` respectively?

File: juju/model.py [hunk 5]
```diff
@@ -3227,16 +3231,16 @@ def make_archive(self, path):
         zf.close()
         return path
 
-    def _check_type(self, path):
+    def _check_type(self, path: str) -> str:
         """Check the path"""
-        s = os.stat(str(path))
+        s = os.stat(path)
```
Comment (james-garner-canonical) [inline juju/model.py:3242]: I think calling `str` here is intending to support `path` to be `str | Path`? Though that's unnecessary since python 3.6 when lots of stdlib stuff started accepting `Path` as an argument.

I see that we already `from pathlib import Path` in this module. But I also see that a lot of these methods are clearly written with the assumption that the argument type is a `str`. So I think your decision to use `str` here is correct

File: juju/utils.py [hunk 6]
```diff
@@ -20,11 +22,11 @@
...
     """Wrapper around asyncio.create_subprocess_exec."""
-    p = await jasyncio.create_subprocess_exec(
+    p = await asyncio.create_subprocess_exec(
         *cmd,
-        stdin=jasyncio.subprocess.PIPE,
-        stdout=jasyncio.subprocess.PIPE,
-        stderr=jasyncio.subprocess.PIPE,
+        stdin=asyncio.subprocess.PIPE,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
```
Comment (james-garner-canonical) [inline juju/utils.py:29]: why the change to not using `jasyncio` here? I notice that it's still imported (line 18)

File: juju/utils.py [hunk 7]
```diff
@@ -93,20 +95,31 @@ class IdQueue:
...
     """
 
-    def __init__(self, maxsize=0):
-        self._queues = defaultdict(partial(jasyncio.Queue, maxsize))
-
-    async def get(self, id_):
+    _queues: dict[int, asyncio.Queue[dict[str, Any] | Exception]]
+
+    def __init__(self):
+        self._queues = defaultdict(asyncio.Queue)
```
Comment (james-garner-canonical) [inline juju/utils.py:101]: Behaviour change? Or is it always called with the default maxsize internally, and we're confident it isn't used externally?

File: juju/utils.py [hunk 8]
```diff
@@ -93,20 +95,31 @@ class IdQueue:
...
         return value
 
-    async def put(self, id_, value):
+    async def put(self, id_: int, value: dict[str, Any]):
         await self._queues[id_].put(value)
 
-    async def put_all(self, value):
+    async def put_all(self, value: Exception):
         for queue in self._queues.values():
             await queue.put(value)
```
Comment (james-garner-canonical) [inline juju/utils.py:124]: Interesting signatures, I won't pry too deeply here, I assume they just reflect how the methods are actually used.

But do you think it's worth annotating `-> None:` (in general) to make it clear that the return type has been considered?

File: juju/model.py [hunk 9]
```diff
@@ -2923,7 +2927,7 @@ async def wait_for_idle(
         status=None,
         wait_for_at_least_units=None,
         wait_for_exact_units=None,
-    ):
+    ) -> None:
```
Comment (james-garner-canonical) [inline juju/model.py:2936]: I guess it might be a lot of work to annotate all the other arguments if they're unusually typed?

I'd be in favour of annotating any that are easy to annotate tho, e.g. if `raise_on_error/blocked` and `wait_for_active` are `bool`, if `timeout`, `idle_period`, `check_freq` are `float` (or `int`?), and I'm guessing that `wait_for_..._units` are `int | None`. Is `status` a `str | None`?

**Review by james-garner-canonical — CHANGES_REQUESTED** [review]
Sorry, I made a lot of comments. You can resolve most without any changes, but I think it would be nice to use something a bit more readable for the `for ... else` one.

File: juju/client/connection.py (hunk 1 above)
Comment (dimaqq) [reply juju/client/connection.py]: I've changed from `Optional` to `x|None` and I guess the `|` was the convenient place for ruff to split the line...

I think I'll make a type alias for the outer dict, that should solve it.

File: juju/client/connection.py (hunk 2 above)
Comment (dimaqq) [reply juju/client/connection.py]: Let me look into this, I think the full type is probably `dict[str, Any]` which is still better than plain `dict`.

Comment (dimaqq) [reply juju/client/connection.py]: P.S. it should complain, it does complain in other contexts... 🤔

File: juju/model.py (hunk 4 above)
Comment (dimaqq) [reply juju/model.py]: pyright seems to imply the type based on the default.
is that not enough?

File: juju/utils.py (hunk 7 above)
Comment (dimaqq) [reply juju/utils.py:101]: yes maxsize is always the default.

File: juju/model.py (hunk 9 above)
Comment (dimaqq) [reply juju/model.py:2936]: There will be a separate PR for .wait_for_idle()

For now I'm sprinkled a few type hints, mostly because I needed them somewhere else.
Does this pass as "it's better than before, let's leave other improvements till later" check?

File: juju/model.py (hunk 4 above)
Comment (james-garner-canonical) [reply juju/model.py]: My concern is that a later reader doesn't necessarily know whether they're not type annotated because the type inferred from the defaults is obviously correct, or because it's like `wait_for_idle` where we're just gradually adding types. Whereas if you add even the obvious annotations, it's clear that the types for this `__init__` method have been fully worked out.

File: juju/model.py (hunk 9 above)
Comment (james-garner-canonical) [reply juju/model.py:2936]: Yeah this makes sense, especially since you're working on a new `wait_for_idle`

File: juju/utils.py (hunk 8 above)
Comment (dimaqq) [reply juju/utils.py:124]: indeed these type hints are for semantically makes sense to call these methods with.

the implementation doesn't care, ofc.

I'm not sure about None return types. I think the annotation is useful only in a limited way -- prevents someone accidentally using the return value, or shows them not to in an IDE.

File: juju/client/connection.py (hunk 2 above)
Comment (dimaqq) [reply juju/client/connection.py]: I've made types for this thing... pushing now.

File: juju/client/connection.py (hunk 3 above)
Comment (dimaqq) [reply juju/client/connection.py:769]: done

File: juju/client/connection.py (hunk 1 above)
Comment (dimaqq) [reply juju/client/connection.py]: done

File: juju/client/facade.py [hunk 10]
```diff
@@ -13,13 +14,22 @@
...
+from typing import Any, Mapping, Sequence
 
 import packaging.version
 import typing_inspect
+from typing_extensions import TypeAlias
 
 from . import codegen
 
+# Plain JSON, what is received from Juju
+_JSON_LEAF: TypeAlias = None | bool | int | float | str
```
Comment (james-garner-canonical) [inline juju/client/facade.py]: Does this work in python 3.8, or do you need to use a Union? We have annotations from the future, but these `|`s aren't in an annotation

File: pyproject.toml [hunk 11]
```diff
@@ -212,6 +212,7 @@ ignore = [
 [tool.pyright]
 # These are tentative
 # include = ["**/*.py"]
-pythonVersion = "3.8" # check no python > 3.8 features are used
-pythonPlatform = "All"
+pythonVersion = "3.10"
```
Comment (james-garner-canonical) [inline pyproject.toml]: Is this ok, don't we still have to run on 20.04?

File: pyproject.toml [hunk 12]
```diff
@@ -212,6 +212,7 @@ ignore = [
...
 typeCheckingMode = "strict"
+useLibraryCodeForTypes = true
+reportGeneralTypeIssues = true
```
Comment (james-garner-canonical) [inline pyproject.toml:218]: What do these do?

File: juju/client/connection.py [hunk 13]
```diff
@@ -26,6 +26,8 @@
 from .facade import _JSON, _RICH_JSON, TypeEncoder
 from .facade_versions import client_facade_versions, known_unsupported_facades
 
+SPECIFIED_FACADES: TypeAlias = dict[str, dict[Literal["versions"], Sequence[int]]]
+
```
Comment (james-garner-canonical) [inline juju/client/connection.py:31]: Looks good, I'm surprised ruff doesn't complain about / fix the case - maybe because it's upper snake? https://docs.astral.sh/ruff/rules/snake-case-type-alias/

File: juju/client/connection.py [hunk 14]
```diff
@@ -536,7 +537,19 @@ async def _do_ping():
...
+    ) -> dict[str, _JSON]: ...
+
+    @overload
+    async def rpc(
+        self, msg: dict[str, _RICH_JSON], encoder: TypeEncoder
+    ) -> dict[str, _JSON]: ...
+
+    async def rpc(
+        self, msg: dict[str, Any], encoder: json.JSONEncoder | None = None
+    ) -> dict[str, Any]:
```
Comment (james-garner-canonical) [inline juju/client/connection.py]: If it's a pain to narrow this definition, let's just leave it with the Any's, but just pointing it out in case you meant to narrow it further

File: pyproject.toml (hunk 12 above)
Comment (dimaqq) [reply pyproject.toml:218]: pyright.useLibraryCodeForTypes [boolean]: This setting is deprecated in favor of python.analysis.useLibraryCodeForTypes. It will be removed at a future time.

python.analysis.useLibraryCodeForTypes [boolean]: Determines whether pyright reads, parses and analyzes library code to extract type information in the absence of type stub files. Type information will typically be incomplete. We recommend using type stubs where possible. The default value for this option is true.

may help with libraries that don't give us types, like pymacaroons


reportGeneralTypeIssues is actuallyu "error" by default, so maybe I don;'t need this setting.


to be honest, I'm just poking a black cat in the black room in the dark when it comes to pyright settings 🐈‍⬛

File: juju/client/connection.py (hunk 13 above)
Comment (dimaqq) [reply juju/client/connection.py:31]: Huh.. interesting...

File: juju/client/connection.py (hunk 14 above)
Comment (dimaqq) [reply juju/client/connection.py]: I don't think it matters. The future me gets to fix it if it's ever a problem.

File: pyproject.toml (hunk 11 above)
Comment (dimaqq) [reply pyproject.toml]: Maybe I poke that bear tomorrow. Maybe, just maybe I can do better 🤔

File: juju/client/facade.py (hunk 10 above)
Comment (dimaqq) [reply juju/client/facade.py]: stringified.

File: juju/model.py (hunk 4 above)
Comment (dimaqq) [reply juju/model.py]: noted.

File: juju/client/connection.py (hunk 13 above)
Comment (dimaqq) [reply juju/client/connection.py:31]: done, thank you!

File: pyproject.toml (hunk 11 above)
Comment (dimaqq) [reply pyproject.toml]: done, thank you for not letting be lazy

**Review by james-garner-canonical — APPROVED** [review]

File: juju/utils.py (hunk 6 above)
Comment (dimaqq) [reply juju/utils.py:29]: this was general cleanup, though I'm not doing the bulk 
 of it, leaving that for #1178
